RFD 225: Teleport Extended Release Support (#59287)
gravitational/teleport: rfd/0225-teleport-extended-support.md

Customers already find our current support matrix confusing, adding in an LTS release is only going to
make the situation worse. Confusion aside, the implications of the answers above would also
require a significant change in our deprecation and compatibility philosophies.

When upgrading to LTS N from LTS N-1, the control plane will first be upgraded to LTS N,
and all agents and clients will still be on LTS N-1. Which means we need to ensure compatibility
for whatever the time span is between two subsequent releases.

In addition to the technical and philosophical challenges that an LTS release would introduce, it would
also have an impact on the tools team and developer experience. We would need to support an additional
number of release branches and build pipelines. 

### Extended Support

While customers might be asking for an LTS release explicitly, what they really want are two
things: stability and extended support. We can achieve these without introducing a Teleport LTS
release by changing our process slightly.

We can reduce the frequency of Teleport major releases from every ~4 months to once a year and cut

> down the number of supported releases from N-2 to N-1. That would extend the support window for major

Reviewer: ```suggestion
down the number of supported releases from 3 to 2. That would extend the support window for major
```

This satisfies customer desire for major releases to be supported for a longer period of time
without impacting our current upgrade process or component compatibility guarantees. However, it
comes at the expense of reducing the frequency at which we can make breaking changes.

As dictated by semantic versioning, a major release should be reserved for making any backward-incompatible
changes to the public API. We've often relied on this four month release cycle to quickly iterate
on APIs that don't work or cause issues for one reason or another. With yearly major releases this
becomes a less viable solution and a stronger emphasis must be put both on backward and forward compatibility. 

Going forward, solutions to problems should attempt to preserve compatibility at all costs and
rely less on all customers adhering to our upgrade procedure. Instead of relying on versions of
clients, and that the Auth server will reject outdated clients we should put a stronger emphasis
on negotiating compatibility. If clients and servers in our control can inform each other about
what features they support in some manner then both can mutually agree on which features and APIs

> can be safely used.

Reviewer 2: > mutually agree on which features and APIs can be safely used

We should note that this is not something that can be done in isolation for a single RPC, any given workflow (which might span multiple RPCs) has to account for the possibility that the remote end might change version and/or bounce between versions between RPCs. We're seeing this already with versioned list RPCs that only support filtering and sorting in some versions, depending on where the fallback happens we might accidentally use a pagination key with the "wrong" RPC version.

To address the stability concerns, all new features *MUST* only be added to the most recent major version.
The previous major version goes into maintenance mode the moment a new major version is released.
In other words, even though we will now support two major releases, N and N-1, new features *MUST* only

> be included in version N. N-1 will continue to receive bug fixes and security patches, but not new features.

Reviewer: this doesn't necessarily pertain to the core of the extended support change, feel free to resolve this comment if you want to avoid derailing the conversation, but I would be so happy if we could start cutting minor release branches off of master and never have to backport another feature.

For example, when we release v19, we would cut `branch/v19.0`. We would only backport bugfixes to `branch/v19.0`, which would actually align with semver. Feature slated for 19.1 would be merged to master, and when we're ready to release 19.1 we would cut `branch/v19.1` from master. Optionally we do the test plan again for 19.1 to continue running through a test plan every 4 months. Once 19.1 is released we can basically retire `branch/v19.0` and only backport bugfixes to `branch/v19.1`

Reviewer 2: I think we rely on master being the staging area for larger features to land in together in a release, so I don't know how much that'd be practical.

### Toolchain Implications

> Extending support of a Teleport means that it may outlive releases of toolchains used to build Teleport.

Reviewer: ```suggestion
Extending support of a Teleport release means that it may outlive releases of toolchains used to build Teleport.
```

Go releases are supported for ~12 months. They release a new major version every ~6 months, and are EOL
when two newer major releases exist.

While Go is typically good about preserving compatibility, new releases do occasionally alter behavior
that require changes to Teleport. We could align our release cycle to the Go release cadence to reduce
the number of times the toolchain would need to be upgraded but that doesn't buy us much.

This isn't an entirely new problem as our current N-2 and N-1 release outlive Go toolchains today.

> We should continue to only bump the Go toolchain on branches _after_ it has gone through a test plan.

Reviewer: after what has gone through a test plan? if we only do a major release once a year will we only be doing the test plan once a year? is the idea that we would upgrade the toolchain used to build v18 releases after we go through the v19 test plan using the latest toolchain?

---
authors: Tim Ross(tim.ross@goteleport.com)
state: draft
---

# RFD 225 - Teleport Extended Release Support

#### Required Approvers

# Required Approvers

> * Engineering: @zmb3 && @russjones && r0mant

Reviewer 3: ```suggestion
* Engineering: @zmb3 && @russjones && @r0mant
```

becomes a less viable solution and a stronger emphasis must be put both on backward and forward compatibility. 

Examples of forward compatibility issues that we've seen in the past:

- Using a boolean instead of a string or an enum - https://protobuf.dev/best-practices/dos-donts/#bad-bools
- Strict validation over a closed set of values prevents adding new ones
- Making new fields mandatory

Examples of backward compatibility issues that we've seen in the past:
- Removing a deprecated API too soon
- Changing the behavior of an existing API
- Altering validation of existing fields of a resource

Going forward, solutions to problems should attempt to preserve compatibility at all costs and
rely less on all customers adhering to our upgrade procedure. Instead of relying on versions of
clients, and that the Auth server will reject outdated clients we should put a stronger emphasis
on negotiating compatibility. If clients and servers in our control can inform each other about
what features they support in some manner then both can mutually agree on which features and APIs
can be safely used. An example of this can be seen with the capabilities exchanged during instance

> heartbeat hello messages: https://github.com/gravitational/teleport/blob/6d4961aed1eef6b360e05facdbcd997fb95593d4/api/proto/teleport/legacy/client/proto/inventory.proto#L136-L181.

Reviewer 4: 👏🏾 

👏🏾 

👏🏾

> To address the stability concerns, all new features *MUST* only be added to the most recent major version.

Reviewer 3: It would be useful to document explicitly what we consider acceptable to go into a release that's in maintenance mode. For example:

- Security fixes.
- Regression fixes.
- Bug fixes that affect customer deployments (unless it's non-trivial?)
- Critical performance fixes that affect customer deployments (unless it's non-trivial?)

I think it will be hard to draw a boundary and we'd probably have to decide on a case by case basis but some starting guidance might be useful.

The previous major version goes into maintenance mode the moment a new major version is released.
In other words, even though we will now support two major releases, N and N-1, new features *MUST* only

> be included in version N. N-1 will continue to receive bug fixes and security patches, but not new features.

Author: I don't think this changes much in terms of when features are available. If users chose stability by staying on the previous major release, then they do not get new features. If users want new features, then they upgrade to current major release and get new features right away. This is by and large no different than today when we only backport a new feature to the current major release.

If a new feature lands in 19.2.0, then it will be available as soon as the cluster is upgraded to 19.2.0 just like today. If I am however running 18.8.0, then I will not get the new feature until I decide that I'm okay with upgrading to >= 19.2.0. 

For Cloud, we can by proceed as we do today and keep all tenants up to date by default. This allows us to make new features available to tenants as soon as we are able to be roll out new versions of Teleport to them. For self-hosted users this means they can get new features as soon as they are willing and able to upgrade to version N of Teleport that introduced it.

IMO we should avoid compromising the stability of the N-1 release by introducing features to it so that those unwilling/unable to upgrade to the N release get them sooner. This strategy is no different than that used other popular enterprise software LTS releases which only include bug fixes and security patches.

> be included in version N. N-1 will continue to receive bug fixes and security patches, but not new features.

Author: New features would land in 18.x minor releases, i.e. 18.1.0, 18.2.0, 18.3.0 while v18 is the current major release. No new features would be backported to v17 during this period. Once 19.0.0 is released, then all new features would land in 19.1.0, 19.2.0, and no new features would be backported to to v18 at that point.

on negotiating compatibility. If clients and servers in our control can inform each other about
what features they support in some manner then both can mutually agree on which features and APIs
can be safely used. An example of this can be seen with the capabilities exchanged during instance
heartbeat hello messages: https://github.com/gravitational/teleport/blob/6d4961aed1eef6b360e05facdbcd997fb95593d4/api/proto/teleport/legacy/client/proto/inventory.proto#L136-L181.

To address the stability concerns, all new features *MUST* only be added to the most recent major version.
The previous major version goes into maintenance mode the moment a new major version is released.
In other words, even though we will now support two major releases, N and N-1, new features *MUST* only
be included in version N. N-1 will continue to receive bug fixes and security patches, but not new features.

What does this mean in practice?

All development will continue on the master branch. New features are only to be backported to
the current release, i.e branch/v18. Bug fixes, security mitigations, will be backported to

> the current release and the stable release, i.e. branch/v18 and branch/v17.

Reviewer 5: I was thinking about big UI changes in the context of this PR. The conclusion I came to is that those changes will need to be relegated from major to minor releases.

For example, Grzegorz is addressing a piece of tech debt blocking him from his work on shared `~/.tsh` between tsh and Connect (#60473). This changes how tsh daemon works but it also changes the UI – for a small percent of users, places which used to refer to the cluster by its name will now show the proxy hostname (which can differ from the cluster name).

Usually, I'd recommend we backport the tsh daemon changes but update the UI only in a major release. This is to avoid a situation where the user performs a minor or patch update and suddenly the app looks totally different.

However, with this RFD in place the UI change would need to wait until August to be released, which is a long time. This change might seem insignificant in the grand scheme of things, maybe better examples would be times when we updated the look and feel of buttons and forms (#43662, #46812) or adjusted how the login form behaves.

I think most of those bigger UI changes can now be relegated to minor updates since otherwise we'd have to wait potentially for months before releasing them.
